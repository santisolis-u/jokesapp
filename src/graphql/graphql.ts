import {ApolloClient, InMemoryCache, gql, HttpLink} from '@apollo/client';
import fetch from 'cross-fetch';

export const client = new ApolloClient({
  link: new HttpLink({
    uri: 'https://icanhazdadjoke.com/graphql',
    fetch,
  }),
  cache: new InMemoryCache(),
});

export const FETCH_DAD_JOKE = gql`
  query {
    joke {
      id
      joke
      permalink
    }
  }
`;
