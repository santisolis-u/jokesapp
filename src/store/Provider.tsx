import React, {useEffect, useState} from 'react';
import {Joke} from '../hooks/useJokes';
import {FavJokesContext} from './store';
import Realm from 'realm';
import {JokeSchema} from '../realm/schema';
interface Props {
  children: JSX.Element;
}

export const FavsProvider: React.FC<Props> = ({children}) => {
  const [favJokes, setFavJokes] = useState<Joke[]>([]);
  const [realm, setRealm] = useState<Realm>();

  useEffect(() => {
    const openRealm = async () => {
      const r = await Realm.open({schema: [JokeSchema]});
      if (r != null && !r.isClosed) {
        setRealm(r);
        const result = r.objects<Joke>('Joke').map(value => {
          return value.toJSON() as Joke;
        });
        setFavJokes(result);
      }
    };
    openRealm();
    return () => {
      if (!!realm && !realm.isClosed) {
        realm.close();
      }
    };
  }, []);

  const addFavJoke = (joke: Joke) => {
    if (!favJokes.includes(joke)) {
      if (!!realm && !realm.isClosed) {
        realm.write(() => {
          realm.create('Joke', joke);
        });
      }
      setFavJokes(favJokes => [...favJokes, joke]);
    }
  };
  const removeFavJoke = (id: string) => {
    if (!!realm && !realm.isClosed) {
      realm.write(() => {
        const joke = realm.objectForPrimaryKey<Joke>('Joke', id);
        realm.delete(joke);
      });
    }
    setFavJokes(favJokes => favJokes.filter(joke => joke.id !== id));
  };
  return (
    <FavJokesContext.Provider value={{favJokes, addFavJoke, removeFavJoke}}>
      {children}
    </FavJokesContext.Provider>
  );
};
