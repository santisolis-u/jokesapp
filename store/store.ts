import {Joke} from '../src/hooks/useJokes';
import React, {FC, ReactChildren, useState} from 'react';
interface FavJokesContextType {
  favJokes: Joke[];
  addFavJoke?: (joke: Joke) => void;
  removeFavJoke?: (id: string) => void;
}

const defaultState = {
  favJokes: [],
};

export const FavJokesContext =
  React.createContext<FavJokesContextType>(defaultState);
