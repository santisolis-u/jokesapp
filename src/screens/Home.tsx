import {
  ActivityIndicator,
  Dimensions,
  Pressable,
  StyleSheet,
  Text,
  View,
} from 'react-native';
import React, {FC, useContext, useLayoutEffect, useState} from 'react';
import Title from '../components/Title';
import Card from '../components/Card';
import Button from '../components/Button';
import Carousel from '../components/Carousel';
import useJokes, {Joke} from '../hooks/useJokes';
import {FavJokesContext} from '../../store/store';
import {NativeStackScreenProps} from '@react-navigation/native-stack';
import {RootStackParamList} from '../../App';
import {JOKE_CARD_COLORS} from '../../utils/Colors';

const {width} = Dimensions.get('window');

type HomeProps = NativeStackScreenProps<RootStackParamList>;

const Home: FC<HomeProps> = ({navigation}) => {
  const {favJokes, addFavJoke} = useContext(FavJokesContext);
  const {jokes, fourJokes, fetchMoreJokes, isLoading} = useJokes();
  const [cardIndex, setCardIndex] = useState(0);

  useLayoutEffect(() => {
    navigation.setOptions({
      headerRight: () => (
        <Pressable onPress={() => navigation.navigate('Saved')}>
          <Text style={styles.saved}>❤️Saved: {favJokes.length} </Text>
        </Pressable>
      ),
    });
  }, [navigation, favJokes]);

  const indexChange = (index: number) => {
    setCardIndex(index % 4);
  };
  console.log('jokes', jokes);
  const handleSaveJoke = (jokeIdx: number) => {
    const favJokeToAdd = fourJokes[jokeIdx];
    const jokeInFav = !!favJokes.find(
      joke => joke.id === fourJokes[jokeIdx].id,
    );

    if (!!addFavJoke && !!favJokeToAdd && !jokeInFav) {
      addFavJoke(favJokeToAdd);
    }
  };
  return (
    <View style={styles.root}>
      <Title text={'Things you can say to annoy designers.'} />
      <View style={styles.carouselContainer}>
        {jokes.length > 0 ? (
          <Carousel
            indexChange={indexChange}
            items={jokes}
            display={isLoading && {opacity: 0.2}}
            cardIndex={cardIndex}
            renderItem={(
              scrollX: any,
              {item, index}: {item: Joke; index: number},
            ) => {
              return (
                <Card
                  isLoading={isLoading}
                  text={item.joke}
                  backgroundColor={JOKE_CARD_COLORS[index % 4]}
                />
              );
            }}
            fetchNextJokes={() => {
              fetchMoreJokes();
            }}
          />
        ) : (
          <View style={styles.loadingContainer}>
            <ActivityIndicator size="large" color={'black'} />
          </View>
        )}
      </View>
      <Button onPress={() => handleSaveJoke(cardIndex)}>
        <Text>Save</Text>
      </Button>
    </View>
  );
};

export default Home;

const styles = StyleSheet.create({
  root: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  carouselContainer: {
    flex: 5,
  },
  carousel: {
    flex: 1,
    backgroundColor: 'pink',
    width: '100%',
  },
  saved: {
    fontFamily: 'OpenSans-Bold',
    fontSize: 14,
  },
  loadingContainer: {
    flex: 1,
    width: width * 0.8,
    alignItems: 'center',
    justifyContent: 'center',
  },
});
