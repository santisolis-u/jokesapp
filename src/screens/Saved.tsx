import {FlatList, StyleSheet, Text, View} from 'react-native';
import React, {useContext} from 'react';
import {FavJokesContext} from '../store/store';
import SavedCards from '../components/SavedCards';

const Saved = () => {
  const {favJokes, removeFavJoke} = useContext(FavJokesContext);
  const handleDelete = (id: string) => {
    if (removeFavJoke) {
      removeFavJoke(id);
    }
  };
  return (
    <View style={styles.savedContainer}>
      {favJokes.length > 0 ? (
        <FlatList
          keyExtractor={element => element.id}
          data={favJokes}
          renderItem={joke => (
            <SavedCards handleDelete={handleDelete} joke={joke.item} />
          )}
          style={styles.flatList}
          showsVerticalScrollIndicator={false}
        />
      ) : (
        <View style={styles.center}>
          <Text style={styles.fallbackText}>Nothing saved</Text>
        </View>
      )}
    </View>
  );
};

export default Saved;

const styles = StyleSheet.create({
  savedContainer: {
    margin: 24,
    // overflow: 'visible',
    flex: 1,
  },
  flatList: {
    // overflow: 'visible',
  },
  center: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  fallbackText: {
    textAlign: 'center',
    fontSize: 32,
    fontFamily: 'OpenSans-Bold',
  },
});
