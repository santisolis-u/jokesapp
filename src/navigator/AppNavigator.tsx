import {StyleSheet, Text, View} from 'react-native';
import React from 'react';
import {NavigationContainer} from '@react-navigation/native';
import Home from '../screens/Home';
import Saved from '../screens/Saved';
import {createNativeStackNavigator} from '@react-navigation/native-stack';

const Stack = createNativeStackNavigator();

export type RootStackParamList = {
  Home: undefined;
  Saved: undefined;
};

const AppNavigator = () => {
  return (
    <NavigationContainer>
      <Stack.Navigator>
        <Stack.Screen name="Home" component={Home} />
        <Stack.Screen
          options={{title: 'Saved jokes'}}
          name="Saved"
          component={Saved}
        />
      </Stack.Navigator>
    </NavigationContainer>
  );
};

export default AppNavigator;

const styles = StyleSheet.create({});
