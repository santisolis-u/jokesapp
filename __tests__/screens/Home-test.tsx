import React from 'react';
import Home from '../../src/screens/Home';
import {MockedProvider} from '@apollo/client/testing';
// Note: test renderer must be required after react-native.
import renderer, {act} from 'react-test-renderer';
import {FETCH_DAD_JOKE} from '../../src/graphql/graphql';
import {fireEvent, render, waitFor} from '@testing-library/react-native';
import {FavJokesContext} from '../../src/store/store';
import AppNavigator from '../../src/navigator/AppNavigator';
import Saved from '../../src/screens/Saved';

const navigation: any = {
  setOptions: jest.fn(),
  navigate: jest.fn(),
};

jest.mock('react-native/Libraries/Animated/NativeAnimatedHelper');

const jokesMocked: any[] = [
  {
    request: {
      query: FETCH_DAD_JOKE,
    },
    result: {
      data: {
        joke: {
          id: '1A',
          joke: 'TESTING JOKE',
          permalink: '404',
        },
      },
    },
  },
];

describe('<Home />', () => {
  test('renders correctly', () => {
    const component = (
      <MockedProvider mocks={jokesMocked}>
        <Home navigation={navigation} />
      </MockedProvider>
    );
    const json = renderer.create(component).toJSON();
    expect(json).toMatchSnapshot();
  });
  test('fires save joke when clicking save button', async () => {
    const mockAddFavJoke = jest.fn();
    const mockRemoveFavJoke = jest.fn();

    const component = (
      <FavJokesContext.Provider
        value={{
          favJokes: [],
          removeFavJoke: mockRemoveFavJoke,
          addFavJoke: mockAddFavJoke,
        }}>
        <MockedProvider mocks={jokesMocked}>
          <Home navigation={navigation} />
        </MockedProvider>
      </FavJokesContext.Provider>
    );
    const {getByText} = await render(component);
    const toClick = await getByText('Save');
    fireEvent.press(toClick);
    await waitFor(() => expect(mockAddFavJoke).toHaveBeenCalledTimes(1));
  });
});
