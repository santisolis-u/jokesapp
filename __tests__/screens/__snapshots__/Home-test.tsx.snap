// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      Array [
        Object {
          "paddingBottom": 42,
          "paddingHorizontal": 46,
          "paddingTop": 42,
        },
        undefined,
      ]
    }
  >
    <Text
      style={
        Object {
          "fontSize": 32,
          "fontWeight": "bold",
          "textAlign": "center",
        }
      }
    >
      Things you can say to annoy designers.
    </Text>
  </View>
  <View
    style={
      Object {
        "flex": 5,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
          "width": 600,
        }
      }
    >
      <ActivityIndicator
        color="black"
        size="large"
      />
    </View>
  </View>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "black",
        "height": 60,
        "justifyContent": "center",
        "marginVertical": 40,
        "width": "80%",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "white",
          "fontFamily": "OpenSans-Regular",
        }
      }
    >
      Save
    </Text>
  </View>
</View>
`;
